wallet: coin selection, don't return results that exceed the max allowed weight

furszy opened the pull request:
Coming from the following comment https://github.com/bitcoin/bitcoin/pull/25729#discussion_r1029324367.

The reason why we are adding hundreds of UTXO from different sources when the target
amount is covered only by one of them is because only SRD returns a usable result.

Context:
In the test, we create 1515 UTXOs with 0.033 BTC each, and 1 UTXO with 50 BTC. Then
perform Coin Selection to fund 49.5 BTC.

As the selection of the 1515 small UTXOs exceeds the max allowed tx size, the
expectation here is to receive a selection result that only contain the big UTXO.
Which is not happening for the following reason:

Knapsack returns a result that exceeds the max allowed transaction size, when
it should return the closest utxo above the target, so we fallback to SRD who
selects coins randomly up until the target is met. So we end up with a selection
result with lot more coins than what is needed.

furszy committed:
coin selection: knapsack, select closest UTXO above target if result exceeds max tx size

The simplest scenario where this is useful is on the 'check_max_weight' unit test
already:

We create 1515 UTXOs with 0.033 BTC each, and 1 UTXO with 50 BTC. Then perform
Coin Selection.

As the selection of the 1515 small UTXOs exceeds the max allowed tx size, the
expectation here is to receive a selection result that only contain the big
UTXO (which is not happening for the reasons stated below).

As knapsack returns a result that exceeds the max allowed transaction size, we
fallback to SRD, which selects coins randomly up until the target is met. So
we end up with a selection result with lot more coins than what is needed.

furszy committed:
coin selection: heap-ify SRD, don't return selection if exceeds max tx weight

Uses a min-effective-value heap, so we can remove the least valuable input/s
while the selected weight exceeds the maximum allowed weight.

Co-authored-by: Murch <murch@murch.one>

furszy committed:
coin selection: BnB, don't return selection if exceeds max allowed tx weight

furszy committed:
wallet: clean post coin selection max weight filter

Now the coin selection algorithms contemplate the
maximum allowed weight internally and return
std::nullopt if their result exceeds it.

furszy committed:
wallet: refactor coin selection algos to return util::Result

so the selection processes can retrieve different errors and not
uninformative std::nullopt

Xekyo commented on src/wallet/coinselection.cpp:
```diff
@@ -105,8 +115,14 @@ std::optional<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_poo
+                    // Set best selection
+                    best_selection = curr_selection;
+                    best_waste = curr_waste;
+                }
```
We should not continue searching a subtree when the selection already exceeds the `max_weight`. Instead of only checking it when we have selected sufficient funds, I would add `curr_selection_weight > max_weight` as a backtrack reason before `else if (curr_value >= selection_target)`:

        if (curr_value + curr_available_value < selection_target || // Cannot possibly reach target with the amount remaining in the curr_available_value.
            curr_value > selection_target + cost_of_change || // Selected value is out of range, go back and try other branch
            (curr_waste > best_waste && is_feerate_high)) { // Don't select things which we know will be more wasteful if the waste is increasing
            backtrack = true;
        } else if (curr_selection_weight > max_weight) { // Exceeding weight for standard tx, cannot find more solutions by adding more inputs
            max_tx_size_exceeded = true; // at least one selection attempt exceeded the max weight
            backtrack = true;
        } else if (curr_value >= selection_target) {       // Selected value is within range

furszy replied:
yeah better 👌🏼 , thanks

murchandamus reviewed (commented):
Concept ACK

furszy commented on src/wallet/spend.cpp:
```diff
@@ -544,27 +544,36 @@ util::Result<SelectionResult> ChooseSelectionResult(const CWallet& wallet, const
     if (auto bnb_result{SelectCoinsBnB(positive_groups, nTargetValue, coin_selection_params.m_cost_of_change)}) {
         results.push_back(*bnb_result);
-    }
+    } else append_error(bnb_result);
```
yeah. Internally the result class contains a variant and the bool operator returns true/false depending on what the variant is holding.

Xekyo replied:
Ah cool, thanks for explaining

furszy commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -997,7 +1005,10 @@ BOOST_AUTO_TEST_CASE(check_max_weight)
             chain, m_args);
 
         BOOST_CHECK(result);
-        BOOST_CHECK(has_coin(result->GetInputSet(), CAmount(50 * COIN)));
+        // Verify that only the 50 BTC UTXO was selected
+        const auto& selection_res = result->GetInputSet();
+        BOOST_CHECK(selection_res.size() == 1);
+        BOOST_CHECK(selection_res.begin()->GetEffectiveValue() == 50 * COIN);
```
hmm that doesn't sound right.

As we set all feerates to 0, the change cost will be 0. So the waste calculation is simply a `waste = inputs total amount - target`.
So we actually will never select the SRD result, because SRD is forced to select the 50 btc UTXO to succeed (otherwise it surpasses the max allowed weight).

In other words, Knapsacks result always return a waste of 0.5 btc, while SRD will return a waste of 0.5 btc + every extra added input that appears on the shuffled vector prior reaching the 50 btc UTXO. So, knapsacks_waste <= srd_waste.

Xekyo replied:
No, the [waste](https://murch.one/posts/waste-metric/) is calculated via `waste = weight_of_inputs × (feerate - long_term_feerate_estimate)` where the LTFRE is 10 ṩ/vB. This means that at a feerate of 0 ṩ/vB, an input set with greater weight results in a lower (“more negative”) waste. I.e. if the SRD solution happens to roll an input set with 500 small inputs plus the 50 ₿ input it would fit in the standard transaction weight and have a better score than the input set with only the 50 ₿ input.

furszy replied:
> the [waste](https://murch.one/posts/waste-metric/) is calculated via waste = weight_of_inputs × (feerate - long_term_feerate_estimate) where the LTFRE is 10 ṩ/vB.

Where is the LTFRE set to 10ṩ/vB?

I'm seeing that it's set to 0:
https://github.com/bitcoin/bitcoin/blob/edc3d1b296e34838d649dc21b8483a52e214932a/src/wallet/test/coinselector_tests.cpp#L971-L972

So inside `GetSelectionWaste` the loop over all the inputs should be returning waste = 0 (waste += input_weight x 0). 
Then below, as `change_cost` is 0 too, we get into:
https://github.com/bitcoin/bitcoin/blob/edc3d1b296e34838d649dc21b8483a52e214932a/src/wallet/coinselection.cpp#L396-L400

Xekyo replied:
Ah, my bad. I did not expect anyone to change LTFRE in tests. :face_with_spiral_eyes:
In that case, the weight of the inputs is irrelevant for the waste score (because `weight × (0 - 0) = 0`), and it will only prefer a changeless input set over one with change. Then it depends on the order in which results are evaluated in `ChooseSelectionResult(…)`. I still don't think this is a good idea.

It would be better to make a test that only uses Knapsack to test the behavior of Knapsack is as expected, instead of depending on the order of evaluation in another function and setting global constants to odd values.

furszy replied:
> In that case, the weight of the inputs is irrelevant for the waste score (because weight × (0 - 0) = 0), and it will prefer an input set with change over one without change (unless excess is also 0). So very likely, the outcome of this test depends on either SRD being unsuccessful and/or that the order in which results are evaluated in ChooseSelectionResult(…) prefers the Knapsack result.

Actually, the result does not depend on the order in which the algorithms are executed. SRD will always succeed and will always have a higher waste score than snapsack (only could fail if after shuffling the vector, the big UTXO is located at last due the exceeded max weight).

The feerates are 0, same as the change cost, so the waste calculation is just a simple: 
```c++
waste = selected_effective_value - target;
```

So, Snapsack result only contains the 50 BTC UTXO (waste = 0,5 btc).
And, SRD result has the 50 BTC UTXO + all the other small UTXOs that appear in the shuffled vector prior to the big UTXO (waste = 0,5 btc + all smaller UTXOs).

Essentially, the best selection that SRD could make is when the big UTXO appears first after shuffling the coins vector, which is equal to the only solution that snapsack returns. All the other solutions that SRD could return have a higher waste than snapsack.

> Since we are explicitly testing the behavior of Knapsack, this should be a test that just runs Knapsack and checks that the Knapsack behavior is expected.

Make a specific test for this sounds good too. But I still think that would be nice to make tests that use the entire coin selection process and check that the expected result is returned (not only checking that any valid result is retrieved): check the inputs size, the value of them, the algorithm used etc.

> Having thought some more about this, setting LTFRE and cost_of_change to 0 basically turns off BnB. It will then only generate solutions that match the target to the satoshi. Definitely feels off that behavior of coin selection algorithms is so fundamentally changed in coinselector_tests.cpp. 🤔

So we have a new working path hehe (probably more suited for a follow-up PR): validate BnB further by reviewing all the test running the entire coin selection process with the feerates and change_cost in 0.
Plus, would be nice to code new tests running coin selection and expecting the best result to be retrieved by BnB.

Xekyo replied:
I think you may be referring to the description before BnB, which only describes the case for changeless.

If you check `GetSelectionWaste(…)`, you'll see that the formula is `waste = input_weight × (feerate - LTFRE) + excess + cost_of_change`. `excess` is the amount of sats dropped to the fees because we overshot the selection target but do not have enough for a change output. It can only be non-zero in changeless solutions. `cost_of_change` is set to zero in changeless solutions, since we do not create change. So, the score will only ever include a non-zero value for one of `excess` or `cost_of_change`. Since both SRD and Knapsack create change, both add `cost_of_change` which is zero here. Since `LTFRE == feerate`, the input weight also gets multiplied with 0. So any input set for both these algorithms will always have a waste score of 0.

Btw, I first described the waste metric.

furszy committed:
test: coin selection, add coverage for SRD

Xekyo commented on src/wallet/coinselection.cpp:
```diff
@@ -188,16 +200,51 @@ util::Result<SelectionResult> SelectCoinsSRD(const std::vector<OutputGroup>& utx
         const OutputGroup& group = utxo_pool.at(i);
         Assume(group.GetSelectionAmount() > 0);
+
+        // If the selection weight exceeds the maximum allowed size, remove the least valuable inputs until we
+        // can insert the new group (while the sum of them are less than the new group amount).
+        weight += group.m_weight;
+        if (weight > max_weight) {
+            max_tx_size_exceeded = true; // mark it in case we don't find any useful result.
+
+            // if the least valuable input is greater, discard the new group.
+            CAmount new_group_amount = group.GetSelectionAmount();
+            if (new_group_amount <= heap.top().GetSelectionAmount()) {
+                weight -= group.m_weight;
+                continue;
+            }
+
+            // Keep popping up inputs while we exceed the max weight and the
+            // arriving 'group' amount is less than the removed inputs.
+            CAmount least_valuable_amounts = 0;
+            do {
+                const OutputGroup& to_remove_group = heap.top();
+                CAmount min_group_amount = to_remove_group.GetSelectionAmount();
+                least_valuable_amounts += min_group_amount;
+                selected_eff_value -= min_group_amount;
+                weight -= to_remove_group.m_weight;
+                heap.pop();
+            } while (!heap.empty() && least_valuable_amounts < new_group_amount && weight > max_weight);
+        }
+
+        // Add group to selection
         selected_eff_value += group.GetSelectionAmount();
-        result.AddInput(group);
+        heap.push(group);
```
It's simpler if you first add the new group to the heap and then use the heap to get below weight:

```suggestion
          
        // Add group to selection
        heap.push(group);
        selected_eff_value += group.GetSelectionAmount();
        weight += group.m_weight;
        
        // If the selection weight exceeds the maximum allowed size, remove the least valuable inputs until we
        // can insert the new group (while the sum of them are less than the new group amount).
        if (weight > max_weight) { 
            max_tx_size_exceeded = true; // mark it in case we don't find any useful result.

            do {
                const OutputGroup& to_remove_group = heap.top();
                CAmount min_group_amount = to_remove_group.GetSelectionAmount();
                selected_eff_value -= min_group_amount;
                weight -= to_remove_group.m_weight;
                heap.pop();
            } while (!heap.empty() && weight > max_weight);
        }
```

furszy replied:
Nice 👌🏼, applied changes and added your co-authorship. Thanks :).

furszy committed:
test: coverage for bnb max weight

Basic positive and negative scenarios

furszy committed:
refactor: coinselector_tests, unify wallet creation code

same lines of code repeated across the entire file over and over.

furszy commented:
Updated with two points:

1) Added coverage for bnb max weight exceeded https://github.com/bitcoin/bitcoin/pull/26720/commits/41fb2c4dd8a15dc0c59ca4d5d0befab707ca8954.

2) Cleaned further the coin selection test f0add1d (noticed by @aureleoules inside [#25806](https://github.com/bitcoin/bitcoin/pull/25806#discussion_r1081460601))

murchandamus reviewed (approved):
crACK w/ nits f0add1d0a7507e5601d7da54f62fe9ab272b3281

furszy commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -940,6 +940,100 @@ BOOST_AUTO_TEST_CASE(effective_value_test)
+            }
+            for (int i = 0; i < 10; i++) { // 10 UTXO --> 20 BTC total --> weight 272 * 10 = 2720 kvb
+                add_coin(available_coins, wallet, CAmount(2 * COIN), CFeeRate(0), 144, false, 0, true);
+            }
```
ah yeah, good catch. All the numbers are in weight units. And yes, inputs are all P2WPKH (all of them in every `coinselector_tests.cpp` test case).

furszy commented on src/wallet/coinselection.cpp:
```diff
@@ -97,6 +99,7 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
     CAmount best_waste = MAX_MONEY;
 
     bool is_feerate_high = utxo_pool.at(0).fee > utxo_pool.at(0).long_term_fee;
+    bool max_tx_size_exceeded = false;
```
fixed

furszy commented on src/wallet/coinselection.cpp:
```diff
@@ -106,6 +109,9 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
             (curr_waste > best_waste && is_feerate_high)) { // Don't select things which we know will be more wasteful if the waste is increasing
             backtrack = true;
+        } else if (curr_selection_weight > max_weight) { // Exceeding weight for standard tx, cannot find more solutions by adding more inputs
+            max_tx_size_exceeded = true; // at least one selection attempt exceeded the max weight
```
done

furszy commented on src/wallet/coinselection.cpp:
```diff
@@ -154,13 +161,14 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
     // Check for solution
     if (best_selection.empty()) {
-        return util::Error();
+        return max_tx_size_exceeded ? ErrorMaxWeightExceeded() : util::Error();
```
done

furszy commented on src/wallet/coinselection.h:
```diff
@@ -384,9 +384,12 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
  * @param[in]  utxo_pool    The positive effective value OutputGroups eligible for selection
  * @param[in]  target_value The target value to select for
+ * @param[in]  rng The randomness source to shuffle coins
+ * @param[in] max_weight The maximum allowed weight for a selection result to be valid
```
done

furszy commented:
Feedback tackled, thanks for the review Xekyo!
Pushed [diff](https://github.com/bitcoin/bitcoin/compare/f0add1d0a7507e5601d7da54f62fe9ab272b3281..7d5a02ecb2af934ea0fa76c9013b2a2fc968b726).

murchandamus reviewed (commented):
reACK 7d5a02ecb2af934ea0fa76c9013b2a2fc968b726

furszy commented:
rebased, conflicts solved.

aureleoules commented on src/wallet/coinselection.cpp:
```diff
@@ -172,9 +173,20 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
+class MinOutputGroupComparator
+{
+public:
+    int operator() (const OutputGroup& group1, const OutputGroup& group2)
```
fd8cf58d736ed9614f46b4f5b3c92f71ff9c46d4
nit
```suggestion
    const int operator() (const OutputGroup& group1, const OutputGroup& group2)
```

furszy replied:
hmm why?
`warning: 'const' type qualifier on return type has no effect`

aureleoules replied:
Whoops I meant `int operator() (const OutputGroup& group1, const OutputGroup& group2) const `

aureleoules reviewed (approved):
ACK 1284223691127e76135a46d251c52416104f0ff1 - I verified the tests are correct and the code looks good to me.

Sidenote, It seems this chunk of code is unused.
<details>
<summary>Unused code</summary>

```diff
diff --git a/src/wallet/coinselection.cpp b/src/wallet/coinselection.cpp
index 92d33bd38..26405d63a 100644
--- a/src/wallet/coinselection.cpp
+++ b/src/wallet/coinselection.cpp
@@ -448,13 +448,6 @@ void OutputGroupTypeMap::Push(const OutputGroup& group, OutputType type, bool in
     }
 }
 
-std::optional<Groups> OutputGroupTypeMap::Find(OutputType type)
-{
-    auto it_by_type = groups_by_type.find(type);
-    if (it_by_type == groups_by_type.end()) return std::nullopt;
-    return it_by_type->second;
-}
-
 CAmount GetSelectionWaste(const std::set<std::shared_ptr<COutput>>& inputs, CAmount change_cost, CAmount target, bool use_effective_value)
 {
     // This function should not be called with empty inputs as that would mean the selection failed
```

</details>

furszy commented on src/wallet/spend.cpp:
```diff
@@ -554,25 +554,34 @@ util::Result<SelectionResult> ChooseSelectionResult(const CAmount& nTargetValue,
     if (auto bnb_result{SelectCoinsBnB(groups.positive_group, nTargetValue, coin_selection_params.m_cost_of_change)}) {
         results.push_back(*bnb_result);
-    }
+    } else append_error(bnb_result);
```
Hmm,  what about something like this https://github.com/furszy/bitcoin-core/commit/0ddd8f4e2672f4826b87b1cfda617a3c15caabea

furszy commented on src/wallet/spend.cpp:
```diff
@@ -563,12 +563,15 @@ util::Result<SelectionResult> ChooseSelectionResult(const CAmount& nTargetValue,
     };
 
+    // Maximum allowed transaction weight
+    int max_weight = MAX_STANDARD_TX_WEIGHT - (coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR);
```
it's the maximum allowed transaction weight, not only for the inputs.

theStack replied:
> it's the maximum allowed transaction weight, not only for the inputs.

I don't think so? The coin selection algos don't have any knowledge about the additional data of the to-be-created tx (static header size + outputs), so they can only work with a weight limit on the inputs.

Xekyo replied:
I’m confused, @aureleoules’s suggestion makes sense to me. If it’s the maximum allowed transaction weight, why is it reduced by the `tx_noinputs_size`? If it’s supposed to be the allowance for inputs, it might additionally need to be reduced by an allowance for the change output?

furszy replied:
> it's the maximum allowed transaction weight, not only for the inputs.
I don't think so? The coin selection algos don't have any knowledge about the additional data of the to-be-created tx (static header size + outputs), so they can only work with a weight limit on the inputs.

oops, I shouldn't answer comments at night. My bad.
it's ok to say that is the max inputs weight. Will rename the variable in the next push.

> If it’s supposed to be the allowance for inputs, it might additionally need to be reduced by an allowance for the change output?

Good point. Need to also deduce the `change_spend_size` for knapsack and SRD.

furszy replied:
pushed

furszy commented:
Thanks aureleoules for the feedback.

> Sidenote, It seems this chunk of code is unused.

That chunk of code is not in master, #27227 already cleaned it.

aureleoules commented:
> That chunk of code is not in master, https://github.com/bitcoin/bitcoin/pull/27227 already cleaned it.

Ah great, I'm not up-to-date with all the recently merged PRs, it's been a while since I reviewed any!

theStack reviewed (commented):
Concept ACK

Changes look good at a first glance :ok_hand:, will do a second round tomorrow.

S3RK commented on src/wallet/coinselection.cpp:
```diff
@@ -13,8 +13,16 @@
+// Common selection error across the algorithms
+static util::Result<SelectionResult> ErrorMaxWeightExceeded()
+{
+    return util::Error{_("The inputs size exceeds the maximum weight. "
```
I wrote #24580 at some point to capture known to me inefficiencies of current coin selection. The scenario I described above is very close to case no.3 in `test_one_big_and_many_small_coins`

furszy replied:
> I haven't written a test for this, so all the following is not verified and just my guess based on the code understanding.
> 
> This is not a regression in this particular PR, but this could lead to misleading errors in a situation when a proper solution exists, but our coin selection failed to find it and only found solution that exceeded weight. Let's imagine following scenario:
> 
> * UTXO pool: two big coins 50BTC, a ton of 0.001BTC.
> * Target amount: 90BTC
> * Proper solution: just take two big coins and create a change output.
> 
> I think none of the coin selection algorithms will be able to **reliably** find the proper solution.
> 
> 1. BnB won't be able to find it, because proper solution includes change.
> 2. Knapsack again optimises for the match, therefore it discards the proper solution in favour of a closer match that exceeds the weight limit.
> 3. SRD misses proper solution with 75% chance, because it checks each coin exactly once. If it skipped at least one of two big coins it will never find the proper solution.

No wait, have you checked the introduced SRD changes? https://github.com/bitcoin/bitcoin/pull/26720/commits/fd8cf58d736ed9614f46b4f5b3c92f71ff9c46d4.

The algorithm now uses a min-effective-value heap; it discards the least valuable input/s
while the selected weight exceeds the maximum allowed weight.

So in your scenario, SRD will always find a solution.

Basically, it adds coins to a heap while the target is not met. So, if the max weight is exceeded, it pops the least valuable coin and adds the new one. Then keeps walking-through the coins vector until (1) target is covered, or (2) fail due an insufficient funds or (3) fail due having enough funds but the max allowed weight was exceeded.

> I think we should a) make the message more explicit about the fact that solution might exist, but we just failed to find one b) propose users to select coins manually as another remedy c) better document the shortcomings of our coin selection.

I think that you are only having in mind part of the process and maybe not contemplating the expanded coin eligibility process.
If no solution was found in the first selection attempt round, we are running the selection algorithms multiple times. Not only once.

An edge case would be that the coins that are needed to create a valid selection are in the last coin eligibility filter, but even in that situation, SRD will find a solution.

Still, let's say that also SRD doesn't return a result for some reason. And we don't find a solution even when the user has enough funds. The solution shouldn't be placed here.
It can be placed at the end of `AutomaticCoinSelection`, because there we know the total available balance (and also the discarded coins, which since acf0119d24c9f8fae825093249a46cd38e4a3a91 we are also contemplating) so we can easily return a "you have enough funds but the wallet coin selection process failed, please select coins manually" instead of the general "Insufficient Funds".

S3RK replied:
ah, yes. You're right, now I see why SRD will always find a solution if it exists. LGTM

furszy commented on src/wallet/coinselection.h:
```diff
@@ -417,10 +418,10 @@ std::optional<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_poo
  * @param[in]  target_value The target value to select for
  * @returns If successful, a SelectionResult, otherwise, std::nullopt
```
pushed

furszy commented:
Thanks all for the feedback :). [Update Diff](https://github.com/bitcoin/bitcoin/compare/ba7970c793f3e8ddb54a02256e7ae52e88e47f08..488dbeab56b2cd62e381883e5fe347b446ae7d2c):

* Renamed `max_weight` to `max_inputs_weight`.
* Deduced change size from the max allowed weight for Knapsack and SRD (not from BnB as it's specialized to not create change).
* Adapted doxygen comment to current sources.

S3RK reviewed (commented):
Almost Code Review ACK 2a475c4
modulo one comment

furszy commented on src/wallet/spend.cpp:
```diff
@@ -563,12 +563,18 @@ util::Result<SelectionResult> ChooseSelectionResult(const CAmount& nTargetValue,
     } else append_error(bnb_result);
 
+    // As Knapsack and SRD can create change, also deduce change weight.
+    max_inputs_weight -= (coin_selection_params.change_spend_size * WITNESS_SCALE_FACTOR);
```
yep thanks, fixed

S3RK commented:
ACK 25ab14712b9d80276016f9fc9bff7fb9c1d09635

The only change is my previous comment addressed about using current variable to determine change output size.

murchandamus reviewed (commented):
reACK 25ab14712b9d80276016f9fc9bff7fb9c1d09635

theStack reviewed (approved):
Code-review ACK 25ab14712b9d80276016f9fc9bff7fb9c1d09635

achow101 commented:
ACK 25ab14712b9d80276016f9fc9bff7fb9c1d09635
